src/app/map/leaflet-map/leaflet-map.component.ts

```diff
@@ -285,6 +255,90 @@ export class LeafletMapComponent implements OnInit, AfterViewInit, OnDestroy {
     }
   }
 
+  public resetMap() {
+    this.fitBounds(); // use default bounds
+  }
+
+  public onLoadStart() { this.loading = true; }
+
+  public onLoadEnd() { this.loading = false; }
+
+  /**
+  * Removes deleted / draws added projects.
+  */
+  private drawMap(deletedApps: Project[], addedApps: Project[]) {
+    let routerProjId = this.router.snapshot.paramMap.get('project');
+
+    // remove deleted apps from list and map
+    deletedApps.forEach(proj => {
+      if (this.markers.has(proj._id)) {
+        this.markerClusterGroup.removeLayer(this.markers.get(proj._id));
+        this.markers.delete(proj._id);
+      }
+    });
+
+    // draw added apps
+    addedApps.forEach((proj) => {
+      if (proj.location && proj.location['coordinates'][1] && proj.location['coordinates'][0]) {
+        const title = `Project: ${proj.name}`;
+        const marker = L.marker(L.latLng(proj.location['coordinates'][1], proj.location['coordinates'][0]), { title: title })
+        .setIcon(markerIconYellow)
+        .on('click', L.Util.bind(this.onMarkerClick, this, proj));
+
+        this.markers.set(proj._id, marker); // save ref to marker
+        this.markerClusterGroup.addLayer(marker); // save to marker clusters group
+
+        // did we navigate to the map with a poject defined?
+        // if so, zoom to the icon and open the popup
+        if (routerProjId && routerProjId === proj._id) {
+          this.selectedProject = proj;
+
+          setTimeout(() => {
+            this.createMarkerPopup(proj, marker, 10);
+          }, 500);
+        }
+      }
+    });
+
+    // set visible apps
+    this.setVisibleDebounced();
+  }
+
+  /**
+   * Sets which apps are currently visible.
+   * Actual function executes no more than once every 250ms.
+   */
+  // tslint:disable-next-line:member-ordering
+  private setVisibleDebounced = _.debounce(this.setVisible, 250);
+
+  private setVisible() {
+    const mapBounds = this.map.getBounds();
+
+    // update visibility for apps with markers only
+    // ie, leave apps without markers 'visible' (as initialized)
+    this.markers.forEach((marker, key) => {
+      const app = this.projects.filter(proj => proj._id === key);
+      if (app[0]) {
+        const markerLatLng = marker.getLatLng();
+        // app is visible if map contains its marker
+        app[0].isVisible = mapBounds.contains(markerLatLng);
+
+        // If there is only one result from the filter
+        // force the popup to auto-display
+        if (this.markers.size === 1 && app[0].isVisible) {
+          if (marker.getPopup()) {
+            marker.openPopup();
+          } else {
+            // create the popup
+            this.createMarkerPopup(app[0], marker, 9);
```

Comment:
Maybe a comment about the '9' is for?

Reply:
sure, I'll add a bit of docs to the createMarkerPopup method as well